src/front/wgsl/mod.rs

```diff
@@ -466,6 +490,126 @@ impl<'a> Error<'a> {
     }
 }
 
+/// Represents the type of the constructor
+///
+/// Vectors, Matrices and Arrays can have partial type information
+/// which later gets inferred from the constructor parameters
+enum ConstructorType {
```

Comment:
there is quite a bit of machinery now associated to constructors. Perhaps it could live in another module?